Transformations: Add standard deviation and variance reducers

selvavm opened the pull request:
**What this PR does / why we need it**: This PR adds reducers for Variance (Population), Variance (Sample), Standard deviation (Population) and Standard deviation (Sample). This reverts the #48997 because #48844 didn't follow review process

<img width="1230" alt="image" src="https://user-images.githubusercontent.com/16543422/170704678-fda7d036-85bd-4d5a-84c4-385a36dc97d3.png">


**Which issue(s) this PR fixes**: #38381, #46313

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**: @torkelo @ryantxu - Please review this

selvavm committed:
Revert "Revert "Transformation: Added variance and standard deviation (#48844)" (#48997)"

This reverts commit a2ebdf2bc2aadc75434f42c6203ac0ab5e13f8d8.

selvavm commented:
Hi, can I get some support on this?

chri2547 commented on docs/sources/panels/calculation-types.md:
```diff
@@ -10,23 +10,27 @@ weight: 1100
 # Reference: Calculation types
 
-Refer to the following list of calculations you can perform in Grafana. You can find these calculations in the **Transform** tab and in the bar gauge, gauge, and stat visualizations.
+You can perform the following calculations in Grafana. Navigate to the **Transform** tab and in the bar gauge, gauge, and stat visualizations.
```
```suggestion
The following table contains a list of calculations you can perform in Grafana. You can find these calculations in the **Transform** tab and in the bar gauge, gauge, and stat visualizations.
```

chri2547 reviewed (approved):
One wording suggestion for you.

selvavm committed:
Update docs/sources/panels/calculation-types.md

Co-authored-by: Christopher Moyer <35463610+chri2547@users.noreply.github.com>

selvavm commented:
@leeoniya - Is it possible for you to support this PR?

selvavm commented:
Hi @torkelo, could you please support in reviewing this PR? 

I tried different alternatives: 1) standard deviation returned as a string but legend is not sortable based on it and legend is not formatted properly; 2) A separate panel with the min, max and standard deviation but the series color mapping is lost and with the repeating panels dashboard, it consume too much space. 

If I am missing a way to get this done, let me know. 

I will be setting up a readily available Grafana cloud instance and hence, I am not willing to build and deploy.

axelavargas commented:
Hey @selvavm 👋🏾, thanks for opening this PR and contributing to Grafana 🚀, I'll take your PR to the Grafana team to discuss if this feature request is something we want to support. As soon as I have news I will let you know.

_ps, sorry for the late reply, we were focused on the last details of Grafana v9.0 :)_

zsofiakomaromigrafana commented:
@josiahg Could you take a look at this feature?

torkelo commented:
@ryantxu @leeoniya any thoughts on the utility / need for this ?

ryantxu commented:
For sure these should not be added as *standard* values that are calculated anytime someone needs min/max :) This can be restructured so they are only calculated when selected.  Perhaps they can require standard to run first, then use the results?

I am also skeptical of the need for both flavors (population/sample) but I can see the need.  All the front-end statistical operations are a bit suspicious since they are often fed pre-agregated data 😬

leeoniya commented:
yep, i don't think we need both population and sample versions, just population would suffice. the only difference in the code here is dividing by `nonNullcount - 1`, which would only have an impact on pretty small datasets [1]. with only a population variant, the named can be shortened as well.

for future:

- it could be useful to offer an option for manually setting the "window" size to all these calcs, in case you're interested in stats for just the last 100 datapoints. this may also take care of the population vs sample distinction by allowing this to be adjusted. 
- all calcs can also work to calculate a "moving avg" or "moving stddev" rather than simply being reducer for the legend.

[1] https://stats.stackexchange.com/questions/3931/intuitive-explanation-for-dividing-by-n-1-when-calculating-standard-deviation/3934#3934

leeoniya commented on packages/grafana-data/src/transformations/fieldReducer.ts:
```diff
@@ -343,6 +377,10 @@ export function doStandardCalcs(field: Field, ignoreNulls: boolean, nullAsZero:
+
+        let _oldMean = calcs.mean;
+        calcs.mean += (currentValue - _oldMean) / calcs.nonNullCount;
+        squareSum += (currentValue - _oldMean) * (currentValue - calcs.mean);
```
can this overflow if you needed to sum 3000 datapoints where each value was in the 1e10 magnitude?

selvavm commented:
Thanks all for supporting this PR

> For sure these should not be added as _standard_ values that are calculated anytime someone needs min/max :) This can be restructured so they are only calculated when selected. Perhaps they can require standard to run first, then use the results?

Thanks @ryantxu for the comments. I added this as standard value because there was no additional looping involved. As suggested, I can make this as a non-standard. How is your preference? If I make this is a non-standard calculation, then, I believe we will need to restructure the `reduce` function to get the calculated values (especially while calculating standard deviation `sqrt(variance)`). I am also unsure how to handle if the variance is not calculated and the standard deviation calculation is requested.

> I am also skeptical of the need for both flavors (population/sample) but I can see the need. All the front-end statistical operations are a bit suspicious since they are often fed pre-agregated data 😬

> the only difference in the code here is dividing by `nonNullcount - 1`, which would only have an impact on pretty small datasets [1].

I also agree to @ryantxu and @leeoniya. For time series, most of the time, the dataset is large and thus, `n-1` does not change the standard deviation much. I added it simply because it was easy to add :).  Please let me know your preference? I can remove the sample calculations.

> , the named can be shortened as well.

Any suggestion? `Variance (P)` and `Standard deviation (P)`?

selvavm committed:
Refactored to make standard deviation and the variance as non-standard calculation
Remove the sample calculations
Shortened the name by renaming Population as p

selvavm committed:
Removed unused variable

selvavm commented:
Hi @ryantxu and @leeoniya. Based on your suggestions, I have done below changes,

- Refactored to make standard deviation and the variance as non-standard calculation
- Refactored to reuse the existing calculation if available
- Removed the variance and standard deviation for the sample dataset
- Shortened the names as Variance (p) and Standard deviation (p)

Let me know if you need me to change something. Thanks.

selvavm committed:
Fixed some lints

selvavm committed:
Revert "Refactored to make standard deviation and the variance as non-standard calculation"

This reverts commit b0b21c559e742eaa6fc71476cc8d166bf849b0b5.

selvavm committed:
Removed refactoring because it needs many changes

selvavm committed:
Merge pull request #1 from selvavm/RefactorFinalRevert

Refactor final revert

selvavm committed:
Added an ! to mean calculation

selvavm commented:
Hi @ryantxu and @leeoniya - Hope you are doing good. Can you please provide your comment on this PR?

selvavm commented:
Hi @axelavargas, @ryantxu - Could you please support in this PR? In my industry we collect lot of time series and we use the standard deviation as an indicator for comparing time series. Hence, not only in my industry but this could be a really helpful indicator for anyone. If there is any risk in merging this but could be done in enterprise version (Azure Grafana), please do let me know.

leeoniya commented:
should we drop "Population" naming from everywhere, since there is only a single variant now, and i think generally it's implied that there is no magic sampling happening when you feed it the full dataset?

if we add a "sample" variant in the future i think we can just make it an additional boolean param passed to the existing `calculateVariance()` function and expose that toggle in the transformer UI.

selvavm committed:
Merge branch 'grafana:main' into main

selvavm committed:
Removed reference to population

selvavm commented:
Hi @leeoniya, thanks for the comments. I like you suggestion of toggle in the future if needed. Based on your comment, I have updated the PR. Please check.
<img width="1230" alt="image" src="https://user-images.githubusercontent.com/16543422/180591272-72c79384-45a2-4ae8-816c-3b6eb350152d.png">

ryantxu commented on packages/grafana-data/src/transformations/fieldReducer.ts:
```diff
@@ -419,6 +435,36 @@ function calculateLastNotNull(field: Field, ignoreNulls: boolean, nullAsZero: bo
+  if (runningNonNullCount > 0) {
+    variance = squareSum / runningNonNullCount;
+  }
+  return { variance: variance };
```
```suggestion
  return { variance: variance, stdDev: Math.sqrt(variance) };
```

and then remove the second function.  This will let both values be calculated at the same time.

selvavm replied:
Hi @ryantxu. I have made change as per your suggestion.

selvavm committed:
Update packages/grafana-data/src/transformations/fieldReducer.ts

Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

selvavm committed:
Combined Variance and StdDev

leeoniya commented:
we've merged these changes in https://github.com/grafana/grafana/pull/52769.

thanks for your contribution, @selvavm, sorry for the long process!

ryantxu commented:
Thank you @selvavm!  this is merged now

selvavm commented:
Thanks @ryantxu, @leeoniya, @axelavargas and @torkelo for supporting on this PR.
